server/middleware/auth.coffee

```diff
@@ -165,6 +167,43 @@ module.exports =
     yield req.logInAsync(user)
     next()
     
+  loginByOAuthProvider: wrap (req, res) ->
+    { provider: providerID, accessToken, code } = req.query
+    unless providerID and (accessToken or code)
+      throw new errors.UnprocessableEntity('Properties "provider" and "accessToken" or "code" required.')
+    
+    if not database.isID(providerID)
+      throw new errors.UnprocessableEntity('"provider" is not a valid id')
+    
+    provider = yield OAuthProvider.findById(providerID)
+    if not provider
+      throw new errors.NotFound('Provider not found.')
+
+    if code and not accessToken
+      { access_token: accessToken } = yield provider.getTokenWithCode(code)
+      if not accessToken
+        throw new errors.UnprocessableEntity('Code lookup failed')
+    
+    userData = yield provider.lookupAccessToken(accessToken)
+    if not userData
+      throw new errors.UnprocessableEntity('Token was invalid')
+    
+    identity = {
+      provider: provider._id
+      id: userData.id
+    }
+    
+    user = yield User.findOne({oAuthIdentities: { $elemMatch: identity }})
+    if not user
+      throw new errors.NotFound('No user with this identity exists')
```

Comment:
Lines 171-198 seem to be the same as api.postUserOAuthIdentity.  Can we pull this out into a common place?